packages/next/client/components/hooks-server.ts

```diff
@@ -3,16 +3,26 @@ import {
   HeadersContext,
   PreviewDataContext,
   CookiesContext,
+  DynamicServerError,
 } from './hooks-server-context'
 
 export function useHeaders() {
+  if (process.env.PHASE === 'phase-production-build') {
```

Comment:
I think this could be solved by setting a server context for static:

```
const StaticGenerationContext = React.createServerContext('StaticGenerationContext', false)
```

```
const isStaticGeneration = useContext(StaticGenerationContext)
if (isStaticGeneration) {
  throw new DynamicServerError('useHeader')
}
```